job_controller/database/db_updater.py

```diff
@@ -174,9 +194,16 @@ def add_detected_run(
             reduction_inputs,
         )
         session = self.session_maker_func()
+        instrument = session.query(Instrument).filter_by(instrument_name=instrument_name).first()
+        if instrument is None:
+            instrument = Instrument(instrument_name=instrument_name)
+            session.add(instrument)
+            session.commit()
```

Comment:
Does this commit need to be called here, as the session will be reused a few lines down? I can never remember the optimal use for the sqlalchemy session since it's been so long.
And if the persistence is needed first, would using a `session.flush` be more appropriate?
This is a question more than a hard suggestion

Reply:
The database generates the id for the instrument, when we commit, we can then get that ID and pass it as part of run.

Reply:
Isn't this slow though and requires two trips?

If the model has the fields, `instrument_id` which is `ForeignKey` and `instrument` which is `relationship(Instrument)` then I believe if you create the instrument, then when you create the run, set the `run.instrument` to be the instrument, the orm will create both objects at the same time, and only require one trip instead of two.

Reply:
Ya see I originally tried this for the function below, but it didn't seem to do it, because at the time that the Run object is created id is None.

Reply:
so if the run  had fields like:
```python
    instrument_id: Mapped[int] = mapped_column(ForeignKey("instruments.id"))
    instrument: Mapped["Instrument"] = relationship("Instrument")

#create an instrument
instrument = Instrument(instrument_name="foo")
# add it to our run
run.instrument = instrument

#save out run
session.add(run) #instrument is automatically added at the same time implicitly 
session.commit()
```

Something like this allows the ORM to create both records at the same time in one transaction.